com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -20,16 +20,16 @@ internal struct VariableKey
         public ulong NetworkObjectId; // the NetworkObjectId of the owning GameObject
         public ushort BehaviourIndex; // the index of the behaviour in this GameObject
         public ushort VariableIndex; // the index of the variable in this NetworkBehaviour
+        public ushort TickWritten; // the network tick at which this variable was set
```

Comment:
This `VariableKey` should implement `IEquatable<VariableKey>` and override the hashcode to avoid boxing and generally for better performance.

Reply:
Nice trick!

Reply:
Can you elaborate where the boxing occurs, currently, and how making the VariableKey equatable would help ? Are you suggesting the Find function to rely on this equatable interface ?

Reply:
I'm slightly terrified that C# is boxing comparisons of `ushort` types

Reply:
I thought that we were using this key in hashsets/dictionary as a key. After looking more closely at the code this seems to not be the case.

The IEquatable interface prevents boxing when doing equality comparisons between two variables of the same type. So `variableKey1.Equals(variableKey2)` allocates if `VariableKey` is a struct type. Usually it's standard practice to have structs which act as a key implement the interface. Here it is not necessary yet but not implementing the interface could burn us in the future if someone uses this in a dictionary/hashset.

Rider can auto generate the interface just implement `Equatable<VariableKey>` and Rider Intelisense will give you then option to generate equality members.

Reply:
This is a pretty old branch. Considering that:
1. This branch uses `(!m_TickIndex || (Entries[i].Key.TickWritten == key.TickWritten)))` (One snapshot cares about tick, the other doesn't)
2. The `snapshot-system-spawn` branch has a plain comparison.
3. Those are not actually used as key.
I'll simply throw a comment for now, and eventually add the IEquatable interface in the spawn branch.